src/rust/engine/src/python.rs

```diff
@@ -270,31 +267,35 @@ impl Key {
   }
 }
 
-///
-/// We wrap PyObject (which cannot be cloned without acquiring the GIL) in an Arc in order to avoid
-/// accessing the Gil in many cases.
-///
+/// We wrap PyObject (aka Py<PyAny>) in an Arc in order to cheaply clone without needing to
+/// access the GIL. It's already cheap to clone `&PyAny` as it simply increases Python's ref
+/// count, but that type only works when we are holding the GIL. While we could clone the
+/// `PyObject` without the GIL, that would be an expensive deep clone.
```

Comment:
I don't think the last sentence is actually true....From `Py<T>`'s docs for `Clone`:

> If the GIL is held this increments self’s reference count. Otherwise this registers the Py<T> instance to have its reference count incremented the next time PyO3 acquires the GIL.

So I'm honestly not sure what the value of wrapping in an `Arc<>` is? Seems like we're adding unnecessary locking. 

`PyObject` (aka `Py<PyAny>`) is already giving us a generic type around arbitrary types. I think we could drop `Value` in a followup and simply use `PyObject`?

Reply:
That entire comment block is more confusing/irrelevant than it is helpful, I think. I'd probably omit it entirely or, as you say, do away with the entire newtype.

> So I'm honestly not sure what the value of wrapping in an Arc<> is?

In this context there is no value, I think.

> It's already cheap to clone `&PyAny` as it simply increases Python's refcount, but that type only works when we are holding the GIL.

Copying and passing around `&PyAny` doesn't involve reference counts, it's a borrowed reference. Having it bound to the gil lifetime allows for cheap and safe working with borrowed references. But this struct needs to *own* a reference to it, which is what `Py<T>` represents.

> While we could clone the `PyObject` without the GIL, that would be an expensive deep clone.

Cloning a `Py<T>` without the GIL just increases a reference count (deferred into the future, when the gil is next acquired). No "deep copying" takes place.

> Seems like we're adding unnecessary locking.

`Arc` doesn't involve locking, just some math on atomic integers. Or I'm misunderstanding your comment.

Reply:
I looked into this while working #13859, and it turns out that wrapping `PyObject` in an `Arc` makes things about 4% faster. Looking at the internals of `Py<T>` though, this doesn't seem too surprising: although `Py<T>` doesn't acquire the GIL, it does acquire a separate global lock to queue increments and decrements, and fully independent atomic increments should always be faster.

That actually points to it maybe being worthwhile to adjust the implementation of `Py<T>` to internally use an `Arc` around the struct that implements `Clone`/`Drop` in terms of `gil::register_incref`/`gil::register_decref`? I can open a ticket if that would be helpful.

Reply:
Thanks for sharing this, that's an interesting finding. I'm hoping in the not too distant future to start work on a `pyo3::v2::prelude` which has simplifications to the object model and improved memory / performance characteristics in some edge cases. This tip is worth remembering.